x/httpproxy/connect_handler.go

```diff
@@ -15,16 +15,43 @@
 package httpproxy
 
 import (
+	"context"
+	"errors"
 	"fmt"
 	"io"
 	"net"
 	"net/http"
+	"strings"
 
 	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/x/config"
 )
 
+type sanitizeErrorDialer struct {
+	transport.StreamDialer
+}
+
+func isCancelledError(err error) bool {
+	if err == nil {
+		return false
+	}
+	// Works around the fact that DNS doesn't return typed errors.
+	return errors.Is(err, context.Canceled) || strings.HasSuffix(err.Error(), "operation was canceled")
+}
+
+func (d *sanitizeErrorDialer) Dial(ctx context.Context, addr string) (transport.StreamConn, error) {
+	conn, err := d.StreamDialer.Dial(ctx, addr)
+	if isCancelledError(err) {
+		return nil, context.Canceled
+	}
+	if err != nil {
+		return nil, errors.New("base dial failed")
+	}
+	return conn, nil
+}
+
 type connectHandler struct {
-	dialer transport.StreamDialer
+	dialer *sanitizeErrorDialer
```

Comment:
Why do we need to return `context.Cancelled`? Will it return a specific HTTP status code?

Reply:
So I can differentiate from a generic base dialer failed. I had issues when using echo + nc because it was closing the request before the response and it took me a while to figure out.